RecoMuon/MuonShowerProducer/plugins/CSCRechitClusterProducer.cc

```diff
@@ -0,0 +1,219 @@
+// -*- C++ -*-
+//
+// Package:    RecoMuon/MuonShowerProducer
+// Class:      MuonShowerProducer
+// 
+/**\class MuonShowerProducer MuonShowerProducer.cc RecoMuon/MuonShowerProducer/plugins/MuonShowerProducer.cc
+
+ Description: [one line class summary]
+
+ Implementation:
+     [Notes on implementation]
+*/
+//
+// Original Author:  Ka Hei Martin Kwok
+//         Created:  Mon, 11 Oct 2021 19:52:09 GMT
+//
+//
+
+
+// system include files
+#include <memory>
+#include <cmath>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include <FWCore/MessageLogger/interface/MessageLogger.h>
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/Common/interface/Ptr.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include <DataFormats/CSCRecHit/interface/CSCRecHit2D.h>
+#include <DataFormats/CSCRecHit/interface/CSCRecHit2DCollection.h>
+#include "DataFormats/MuonReco/interface/MuonShowerCSCCluster.h"
+#include "Geometry/CSCGeometry/interface/CSCGeometry.h"
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "DataFormats/MuonDetId/interface/CSCDetId.h"
+
+#include "fastjet/JetDefinition.hh"
+#include "fastjet/ClusterSequence.hh"
+#include "fastjet/PseudoJet.hh"
+//
+// class declaration
+//
+typedef edm::Ref<CSCRecHit2DCollection>  rechitRef;
+typedef std::vector<rechitRef>  rechits;
+
+class CSCRechitClusterProducer : public edm::stream::EDProducer<> {
```

Comment:
From what I understand in the `edm::Ref` twiki, it cannot be use polymorphically. So I'm not sure how this can be turned into a base class. 

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMRef